go/vt/schemadiff/capability.go

```diff
@@ -0,0 +1,204 @@
+	switch opt := alterOption.(type) {
+	case *sqlparser.ChangeColumn:
+		// We do not support INSTANT for renaming a column (ALTER TABLE ...CHANGE) because:
+		// 1. We discourage column rename
+		// 2. We do not produce CHANGE statements in declarative diff
+		// 3. The success of the operation depends on whether the column is referenced by a foreign key
+		//    in another table. Which is a bit too much to compute here.
+		return false, nil
+	case *sqlparser.AddColumns:
+		if opt.First || opt.After != nil {
+			// not a "last" column. Only supported as of 8.0.29
+			return capableOf(capabilities.InstantAddDropColumnFlavorCapability)
+		}
+		// Adding a *last* column is supported in 8.0
+		return capableOf(capabilities.InstantAddLastColumnFlavorCapability)
+	case *sqlparser.DropColumn:
+		// not supported in COMPRESSED tables
+		if opt := findTableOption("ROW_FORMAT"); opt != nil {
+			if strings.EqualFold(opt.String, "COMPRESSED") {
+				return false, nil
+			}
+		}
+		if isVirtualColumn(opt.Name.Name.String()) {
+			// supported by all 8.0 versions
+			return capableOf(capabilities.InstantAddDropVirtualColumnFlavorCapability)
+		}
+		return capableOf(capabilities.InstantAddDropColumnFlavorCapability)
+	case *sqlparser.ModifyColumn:
+		if col := findColumn(opt.NewColDefinition.Name.String()); col != nil {
+			// Check if only diff is change of default
+			// we temporarily remove the DEFAULT expression (if any) from both
+			// table and ALTER statement, and compare the columns: if they're otherwise equal,
+			// then the only change can be an addition/change/removal of DEFAULT, which
+			// is instant-table.
+			tableColDefinition := colStringStrippedDown(col, true, false)
+			newColDefinition := colStringStrippedDown(opt.NewColDefinition, true, false)
+			if tableColDefinition == newColDefinition {
+				return capableOf(capabilities.InstantChangeColumnDefaultFlavorCapability)
+			}
+			// Check if:
+			// 1. this an ENUM/SET
+			// 2. and the change is to append values to the end of the list
+			// 3. and the number of added values does not increase the storage size for the enum/set
+			// 4. while still not caring about a change in the default value
+			if len(col.Type.EnumValues) > 0 && len(opt.NewColDefinition.Type.EnumValues) > 0 {
+				// both are enum or set
+				if !hasPrefix(opt.NewColDefinition.Type.EnumValues, col.Type.EnumValues) {
+					return false, nil
+				}
+				// we know the new column definition is identical to, or extends, the old definition.
+				// Now validate storage:
+				if strings.EqualFold(col.Type.Type, "enum") {
+					if len(col.Type.EnumValues) <= 255 && len(opt.NewColDefinition.Type.EnumValues) > 255 {
+						// this increases the SET storage size (1 byte for up to 8 values, 2 bytes beyond)
+						return false, nil
+					}
+				}
+				if strings.EqualFold(col.Type.Type, "set") {
+					if (len(col.Type.EnumValues)+7)/8 != (len(opt.NewColDefinition.Type.EnumValues)+7)/8 {
```

Comment:
We don't need to perform any math if we're checking for equality here, do we? i.e., I don't think we need to do more than:
```
if len(col.Type.EnumValues) != len(opt.NewColDefinition.Type.EnumValues) {
```

Maybe that's not what we want to do though either? I'm assuming it's fine if the new value is less than the existing one. I would think we only care if the new value would be larger / require more space:
```
if (len(col.Type.EnumValues)+7)/8 < (len(opt.NewColDefinition.Type.EnumValues)+7)/8 {
```

Reply:
Unfortunately both assumptions are incorrect. The condition for allowing `ALGORITHM=INSTANT` is that the storage space for the `SET/ENUM` value is identical.

Here's how storage for `6` values is the same as the storage for `7` values, and `INSTANT` is allowed:
```sql
> create table t(id int, c1 set('a', 'b', 'c', 'd', 'e', 'f'), primary key(id));
Query OK, 0 rows affected (0.01 sec)

> alter table t modify column c1 set('a', 'b', 'c', 'd', 'e', 'f', 'g'), algorithm=instant;
Query OK, 0 rows affected (0.01 sec)
Records: 0  Duplicates: 0  Warnings: 0
```

Here's how the storage for `9` (2 bytes) values reduced to `8` values (1 byte) disables `INSTANT`:

```sql
> create table t(id int, c1 set('a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i'), primary key(id));
Query OK, 0 rows affected (0.01 sec)

> alter table t modify column c1 set('a', 'b', 'c', 'd', 'e', 'f', 'g', 'h'), algorithm=instant;
ERROR 1846 (0A000): ALGORITHM=INSTANT is not supported. Reason: Need to rebuild the table to change column type. Try ALGORITHM=COPY/INPLACE.
```